### ISSUE TITLE
Python 3 compatibility

### ISSUE THREAD

**Issue by GreyCat** [kaitai-io/kaitai_struct_python_runtime#6]
Current situation is available at http://kaitai.io/ci

As of now:

* 53 tests total
* Python 2: 1 test fails (SwitchManualEnum)
* Python 3: 17 errors, 1 failure

**Comment by dgladkov** [comment]
I see columns "python", "python2" and "python3" in the CI results table, that doesn't seem right. I don't understand CI configuration yet, so maybe I'm missing something.

**Comment by GreyCat** [comment]
There are 2 kinds of specs:

* Expression translation specs - i.e. taking KS internal expression language expression and translate it into the target language expression
* .ksy compilation and running - taking .ksy + some data in that format

So far, Python 2 and Python 3 expression output are exactly the same - and, as far as I understand, the general idea is to try keeping it that way. The output of compiler is the same both both versions of Python, but the test run results differ.

**Comment by GreyCat** [comment]
Several last commits improved the situation a lot, now it's only one major problem remains with Python 3 (ProcessXor4Const and ProcessXor4Value tests), i.e. this place:

```python
    @staticmethod
    def process_xor_many(data, key):
        r = array.array('B', data)
        ki = 0
        kl = len(key)
        for i in range(len(r)):
            r[i] ^= ord(key[ki])
            ki += 1
            if ki >= kl:
                ki = 0
        return r.tostring()
```

The problem I see is that:

* In Python 2, `key` is actually a `str`, and `key[ki]` returns another 1-character long `str`, so we need to do `ord` to get byte value as integer to do XORing
* In Python 3, `key` is `bytes`, thus `key[ki]` returns integer byte value right away => no `ord` should be performed over it

@dgladkov: any ideas how to make this code compatible with both Python 2 and 3?

**Comment by dgladkov** [comment]
Changed `array.array` to `bytearray`, also used `bytearray` to store xor key to avoid problem above. Using `bytearray` also saves us from following deprecation warning:
```
DeprecationWarning: tostring() is deprecated. Use tobytes() instead.
```

**Comment by dgladkov** [comment]
@GreyCat can you check this solution? I'm still figuring out how to run test suite on my machine.

**Comment by GreyCat** [comment]
Yeah, sure, I'm trying it now.
Running a test suite should be easy - i.e. just do a checkout of [the main project](https://github.com/kaitai-io/kaitai_struct), and then:

```shell
cd tests
./build-compiler
./build-formats
./run-python
```

`build-compiler` will require [sbt](http://www.scala-sbt.org/), which in turn requires you to have a Java runtime environment installed. That's all, all other dependencies would be downloaded automatically on the first build.

**Comment by GreyCat** [comment]
@dgladkov Your changes work flawlessly, thanks! I'll fix the remaining spec now and that goes off to CI.

**Comment by GreyCat** [comment]
Basically, the only thing that's different from 2 now is a bunch of warnings:

```
tests/compiled/python/switch_manual_int_else.py:23: ResourceWarning: unclosed file <_io.FileIO name='src/enum_0.bin' mode='rb' closefd=True>
  self._io = _io
tests/compiled/python/switch_manual_int_else.py:23: ResourceWarning: unclosed file <_io.FileIO name='src/if_struct.bin' mode='rb' closefd=True>
  self._io = _io
tests/compiled/python/switch_manual_int_else.py:23: ResourceWarning: unclosed file <_io.FileIO name='src/term_strz.bin' mode='rb' closefd=True>
  self._io = _io
```

Shall we ignore them for now, or can we do something about them?

**Comment by dgladkov** [comment]
This looks like a problem in those tests, they should close KaitaiStream by calling close method.

**Comment by GreyCat** [comment]
Well, in real world app, it could be a problem, but in a test environment, adding manual `r._io.close` is just bothersome and adds more boilerplate to the tests. Is there a way to disable these warnings and/or add automatic file close call on test exit?

**Comment by dgladkov** [comment]
@GreyCat created a PR, now you can use KaitaiStruct as context manager. To solve this issue in tests you have to either call close() on KaitaiStruct at the end of the test or use KaitaiStruct as context manager. Suppressing warnings doesn't seem like a good idea to me.

**Comment by GreyCat** [comment]
Makes sense, thanks! So, am I right that it all boils down to adjusting all tests from:

```python
        r = HelloWorld.from_file('src/fixed_struct.bin')
        self.assertEqual(r.one, 0x50)
```

to

```python
        with HelloWorld.from_file('src/fixed_struct.bin') as r:
            self.assertEqual(r.one, 0x50)
```

?

**Comment by dgladkov** [comment]
That's right:)

**Comment by GreyCat** [comment]
Ok, I'll try to write a script to do that conversion now...

**Comment by GreyCat** [comment]
Here we go: https://github.com/kaitai-io/kaitai_struct_tests/commit/6b16594720ffa19111628e2eb9d42ea60118261f

Looks like it's all good and done, except for [xor4 that got hit in collateral damage](https://github.com/kaitai-io/kaitai_struct_python_runtime/commit/04289cc13d6501153cab61e875d9c4e2ccdaa5d8#commitcomment-19375931) ;)

**Comment by GreyCat** [comment]
And another small question: how to convert https://github.com/kaitai-io/kaitai_struct_tests/blob/master/spec/python/test_eof_exception_bytes.py properly to the `with` style?

**Comment by dgladkov** [comment]
Looks ugly, but should work

```python
import unittest

from eof_exception_bytes import EofExceptionBytes

class TestEofExceptionBytes(unittest.TestCase):
    def test_eof_exception_bytes(self):
        with self.assertRaises(EOFError):
            with EofExceptionBytes.from_file("src/term_strz.bin"):
                pass
```

Another good thing about context managers is that exit is guaranteed to run even after exceptions inside context manager:

```
In [28]: class TestCM:
    ...:     def __enter__(self):
    ...:         print('Enter')
    ...:         return self
    ...:     def __exit__(self, *args, **kwargs):
    ...:         print('Exit')
    ...:     def raiseException(self):
    ...:         raise Exception
    ...:     

In [29]: with TestCM() as cm:
    ...:     cm.raiseException()
    ...:     
    ...: 
Enter
Exit
---------------------------------------------------------------------------
Exception                                 Traceback (most recent call last)
<ipython-input-31-972654aa21a9> in <module>()
      1 with TestCM() as cm:
----> 2     cm.raiseException()
      3 

<ipython-input-28-05988057d5c9> in raiseException(self)
      6         print('Exit')
      7     def raiseException(self):
----> 8         raise Exception
      9 

Exception:
```

**Comment by dgladkov** [comment]
Wait, I just realized that EOFError will be thrown by `open` and not our code, so you don't need to clean that up because descriptor should not be created anyway.

**Comment by GreyCat** [comment]
@dgladkov Um, it is thrown exactly by our code in this test. The idea is to test this:

```python
    def read_bytes(self, n):
        r = self._io.read(n)
        if len(r) < n:
            raise EOFError(
                "requested %d bytes, but got only %d bytes", n, len(r)
            )
        return r
```

And this will be called in object's constuctor, thus ruining construction of object.

Unfortunately, I've added your suggestion and it's still complaining about unclosed file:

```
/usr/lib/python3.5/traceback.py:282: ResourceWarning: unclosed file <_io.FileIO name='src/term_strz.bin' mode='rb' closefd=True>
  self._line = linecache.getline(self.filename, self.lineno).strip()
```

Is it something with my environment? Does it work for you?

**Comment by dgladkov** [comment]
@GreyCat OK, I was wrong, `__exit__` will not actually be executed if class init throws exception. This can also happen to user if he makes mistake in his ksy file or tries to read wrong or corrupted file, so we should handle it properly by KaitaiStruct class. This issue is resolved in @82f11571acc21564284d73406ce73889c6336da8, no change in test suite is required.

**Comment by GreyCat** [comment]
Great job, thanks! I guess that concludes Python 3 vs Python 2 compatibility issue - well, to be honest, it's much more that was expected :)
